File: Data Structures & Algorithms/linked-list-cycle-detection/submission-2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    bool hasCycle(ListNode* head) {
        //fast slow pointer
        /*
            1   2   3   4
                        s
                        f
        */
        ListNode* fast = head && head->next ? head->next->next : nullptr;
        ListNode* slow = head;

        while(slow && fast){
            if(slow == fast){
                return true;
            }

            slow = slow->next;
            fast = fast->next && fast->next->next ? fast->next->next : nullptr;
        }
        return false;
    }
};

File: Data Structures & Algorithms/reverse-a-linked-list/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, old_head: Optional[ListNode]) -> Optional[ListNode]:
        self.head, self.tail = None, None

        def insert(val:int) -> None:
            nnode = ListNode(val)

            if self.head == self.tail == None:
                self.head = self.tail = nnode
            
            else:
                self.tail.next = nnode
                self.tail = nnode
        
        def reverse(node: Optional[ListNode]):
            if not node:
                return
            
            reverse(node.next)
            insert(node.val)
        
        reverse(old_head)
        return self.head
